lib/teleterm/clusters/cluster_access_requests.go

```diff
@@ -210,3 +265,48 @@ func (c *Cluster) AssumeRole(ctx context.Context, req *api.AssumeRoleRequest) er
 
 	return nil
 }
+
+func getResourceDetails(ctx context.Context, req types.AccessRequest, clt auth.ClientI) (map[string]ResourceDetails, error) {
```

Comment:
> The `getResourceDetails` method is mostly copied from the web PR linked above, but since that lives in `e` only (and slightly different types), I've opted to copy this method due to tsh not having an `e` equivalent and Connect needs to support any type of cluster

I don't think copying code here is a good idea. We already have a lot of UI code copied between the Web UI and Connect. I feel like long term this will make it even harder to introduce changes to access requests in Connect because:

1) Even more code will need to be copied between those two
2) The person changing the code in the Web UI might not even realize that Connect will require changes because everything will compile just fine after making changes only in the Web UI

I don't think we'll ever want to intentionally introduce a change which will make Connect lag behind the Web UI in an area which we already worked on.

---

I see that both places use `types.ResourceID` and `services.GetResourcesByResourceIDs`. The only difference is the return type and how the client is obtained. How about the following:

1. `getResourceDetails` is extracted to `services.GetResourceDetails`.
2. In `api/types/resource.go`, we add `types.ResourceDetails`. The Web UI and Connect will recast this to their own types.
3. When it comes to parameterizing how the client is fetched: I don't know how this is usually done in Go, maybe @zmb3 has some ideas, but what about making `services.GetResourceDetails` accept a function as the last argument. The function would receive a cluster name as input and return a client.

This would make both places share the business logic and we'd only be left with the tedious part of changing the types.

Reply:
In general, I feel like we have to make a conscious effort to share code with the Web UI or tsh.

Otherwise we will always repeat a scenario where someone more experienced with the codebase makes a change in tsh / web backend code and then the less experienced Connect team, afraid of making project-wide changes, comes in and implements the change by copying the implementation from tsh / web backend.

At least for me it always felt like the easier choice but long term I don't think it's good.

Reply:
I don't think we'd have to parameterize the client fetching, that can be handled individually just like the type casting will be, imo. I agree with the points you raise and will abstract these out. The individual methods used in the web/Connect codebase will just fetch a client and cast the types, and the rest will be handled in the `services.GetResourceDetails` method. Thanks!